h/static/scripts/annotation-ui.js

```diff
@@ -129,11 +129,60 @@ function excludeAnnotations(current, annotations) {
   });
 }
 
+function findByID(annotations, id) {
+  return annotations.find(function (annot) {
+    return annot.id === id;
+  });
+}
+
+function findByTag(annotations, tag) {
+  return annotations.find(function (annot) {
+    return annot.$$tag === tag;
+  });
+}
+
 function annotationsReducer(state, action) {
   switch (action.type) {
   case types.ADD_ANNOTATIONS:
-    return Object.assign({}, state,
-        {annotations: state.annotations.concat(action.annotations)});
+    {
+      var updatedIDs = {};
+      var updatedTags = {};
+
+      var added = [];
+      var unchanged = [];
+      var updated = [];
+
+      action.annotations.forEach(function (annot) {
+        var existing = findByID(state.annotations, annot.id);
```

Comment:
It's worth noting that this is now `O(N * M)` where N is the number of annotations loaded and M is the number of annotations being loaded. I guess this hasn't caused any problems but it seems like at some point maintaining maps from id/$$tag to the annotations might be a good idea.